kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,440 @@
+{-|
+Copyright   : (c) Runtime Verification, 2020
+License     : NCSA
+
+-}
+
+module Kore.Step.Implication
+    ( OneAllPath (..)
+    , Implication (..)
+    , freeVariablesLeft
+    , freeVariablesRight
+    , implication
+    , substitute
+    , assertRefreshed
+    , refreshExistentials
+    , applySubstitution
+    , termToExistentials
+    , mkGoal
+    , forgetSimplified
+    , parseRightHandSide
+    , implicationToTerm
+    ) where
+
+import Prelude.Kore
+
+import Control.DeepSeq ( NFData )
+import Control.Error.Util ( hush )
+import Control.Monad.State.Strict ( evalState )
+import qualified Control.Monad.State.Strict as State
+import qualified Data.Default as Default
+import Data.Map.Strict ( Map )
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified GHC.Generics as GHC
+import qualified Generics.SOP as SOP
+
+import qualified Kore.Attribute.Axiom as Attribute
+import Kore.Attribute.Pattern.FreeVariables
+    ( FreeVariables
+    , HasFreeVariables (..)
+    )
+import qualified Kore.Attribute.Pattern.FreeVariables as FreeVariables
+import Kore.Debug
+import qualified Kore.Internal.Condition as Condition
+import Kore.Internal.OrPattern ( OrPattern )
+import qualified Kore.Internal.OrPattern as OrPattern
+import Kore.Internal.Pattern ( Pattern )
+import qualified Kore.Internal.Pattern as Pattern
+import qualified Kore.Internal.Predicate as Predicate
+import Kore.Internal.Substitution ( Substitution )
+import qualified Kore.Internal.Substitution as Substitution
+import Kore.Internal.Symbol ( Symbol )
+import Kore.Internal.TermLike
+    ( ElementVariable
+    , InternalVariable
+    , Modality
+    , SomeVariable
+    , SomeVariableName (..)
+    , TermLike
+    , Variable (..)
+    , VariableName
+    )
+import qualified Kore.Internal.TermLike as TermLike
+import Kore.Rewriting.RewritingVariable
+    ( RewritingVariableName
+    , getRewritingTerm
+    , resetConfigVariable
+    )
+import Kore.Rewriting.UnifyingRule ( UnifyingRule (..) )
+import Kore.TopBottom ( TopBottom (..) )
+import Kore.Unparser ( Unparse (..) )
+import Kore.Variables.Fresh ( refreshVariables )
+
+import Pretty ( Pretty (..) )
+import qualified Pretty
+
+-- Modalities
+data OneAllPath = OnePath | AllPath 
```

Comment:
We shouldn't define the modalities here; that would make the module dependencies awkward.